### PR TITLE
Reduce the work to add a new expr: rewrite SubstituteInExpr with OptOutMutator

### PR DESCRIPTION
**Pull request by zasdfgbnm**
stacked on https://github.com/csarofeen/pytorch/pull/2199

### CODE REVIEW

**Review by csarofeen — APPROVED** [review]
LGTM

File: torch/csrc/jit/codegen/cuda/ir_utils.cpp [hunk 1]
```diff
@@ -155,397 +155,31 @@ std::vector<int> normalizeOld2New(
...
-
-    expr_ = IrBuilder::create<ShiftOp>(
-        shift_expr->container(),
-        out,
-        in,
-        shift_expr->offsets(),
-        shift_expr->padWidth());
-  }
+ protected:
+  virtual void removeExpr(IrContainer*, Expr*) const override {}
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/ir_utils.cpp:171]: Nice trick!

File: torch/csrc/jit/codegen/cuda/ir_utils.cpp [hunk 2]
```diff
@@ -155,397 +155,31 @@ std::vector<int> normalizeOld2New(
...
-        ? substitute_->as<TensorView>()
-        : mma_expr->out();
-    auto init = reference_->sameAs(mma_expr->init())
-        ? substitute_->as<TensorView>()
-        : mma_expr->init();
-    expr_ = IrBuilder::create<MmaOp>(
-        mma_expr->container(), out, in_a, in_b, init, mma_expr->options());
+ private:
+  explicit SubstituteInExpr(Val* reference, Val* substitute) {
+    mutations[reference] = substitute;
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/ir_utils.cpp]: nit: rename this to `mutations_`

**Review by naoyam — APPROVED** [review]

File: torch/csrc/jit/codegen/cuda/ir_utils.cpp (hunk 2 above)
Comment (zasdfgbnm) [reply torch/csrc/jit/codegen/cuda/ir_utils.cpp]: renamed

**Pull request merged**